constraint/pkg/client/drivers/k8scel/transform/cel_snippets_test.go

```diff
@@ -0,0 +1,733 @@
+package transform
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"testing"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	v1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	rSchema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/admission/plugin/cel"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/matchconditions"
+	"k8s.io/apiserver/pkg/cel/environment"
+	"k8s.io/utils/ptr"
+)
+
+func shouldMatch(shouldMatch, shouldErr bool, matchResult matchconditions.MatchResult) error {
+	hadErr := false
+	if matchResult.Error != nil {
+		hadErr = true
+		if !shouldErr {
+			return fmt.Errorf("failed to evaluate matcher: %w", matchResult.Error)
+		}
+	}
+	if matchResult.Matches != shouldMatch {
+		return fmt.Errorf("got matches = %v; wanted %v", matchResult.Matches, shouldMatch)
+	}
+	if !hadErr && shouldErr {
+		return errors.New("expected error, not none")
+	}
+	return nil
+}
+
+type kindMatchEntry struct {
+	Kinds  []string `json:"kinds,omitempty"`
+	Groups []string `json:"apiGroups,omitempty"`
+}
+
+type kindMatcher []kindMatchEntry
+
+func (km *kindMatcher) ToUnstructured() (interface{}, error) {
+	bytes, err := json.Marshal(km)
+	if err != nil {
+		return nil, err
+	}
+	var ret interface{}
+	if err := json.Unmarshal(bytes, &ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+type nsMatcher []string
+
+func (nm *nsMatcher) ToUnstructured() (interface{}, error) {
+	bytes, err := json.Marshal(nm)
+	if err != nil {
+		return nil, err
+	}
+	var ret interface{}
+	if err := json.Unmarshal(bytes, &ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+func TestMatchKinds(t *testing.T) {
```

Comment:
is it worth having a test that includes a mix of matchers ?

Reply:
it is not worth it